docs - embedding a chart (#78856)
metabase/metabase: docs/embedding/query-builder.md

---
title: Embed the query builder
summary: "Embed Metabase's visual query builder or SQL editor in your app, so people can build and save their own questions."
---

# Embed the query builder

{% include plans-blockquote.html feature="Embedding the query builder" convert_pro_link_to_embedding=true %}

> You can embed one of Metabase's editors so people can build questions from scratch.

Author: ```suggestion
You can embed one of Metabase's query editors so people can build questions from scratch.
```

- [Visual query builder](#embed-the-visual-query-builder)
- [SQL editor](#embed-the-sql-editor)

> Both editors require SSO, because in order to run a new query, Metabase has to know who's asking, so it can work out which data they're allowed to see.

Author: ```suggestion
Both query editors require SSO, because in order to run a new query, Metabase has to know who's asking, so it can work out which data they're allowed to see.
```

![Query builder](./images/embedded-query-builder.png)

As a web component:

```html
<metabase-question question-id="new"></metabase-question>
```

With the SDK:

```tsx
{% include_file "{{ dirname }}/sdk/snippets/questions/new-question.tsx" %}
```

To narrow down what people can start from, list the entity types you want in the data picker with the `entity-types` attribute (web component) or the `entityTypes` prop (SDK). For example, `entity-types="['model']"` limits the picker to [models](../data-modeling/models.md).

## Embed the SQL editor

![Embedded SQL editor](./images/embedded-sql-editor.png)

To let people write native SQL, use `new-native` as the question ID.

As a web component:

```html
<metabase-question question-id="new-native"></metabase-question>
```

With the SDK:

```tsx
{% include_file "{{ dirname }}/sdk/snippets/questions/new-native-question.tsx" %}
```

Everyone still queries through their own Metabase account, so people can only run SQL against databases their groups have permission to query. See [data permissions](../permissions/data.md).

## Let people save questions

### Saving with web components

> With a web component, turn saving on with `is-save-enabled="true"`, and set the collection that new questions land in with `target-collection`:

Author: ```suggestion
With a web component, turn saving on with `is-save-enabled="true"`, and set the collection for new questions to land in with `target-collection`:
```

---
title: Embed the query builder
summary: "Embed Metabase's visual query builder or SQL editor in your app, so people can build and save their own questions."
---

# Embed the query builder

{% include plans-blockquote.html feature="Embedding the query builder" convert_pro_link_to_embedding=true %}

You can embed one of Metabase's query editors so people can build questions from scratch.

- [Visual query builder](#embed-the-visual-query-builder)
- [SQL editor](#embed-the-sql-editor)

> Both query editors require SSO, because in order to run a new query, Metabase has to know who's asking, so it can work out which data they're allowed to see.

Reviewer: Well technically it requires user login, SSO is just the only means of providing it. Can we link maybe the overview article where we talk about guest and SSO?

To embed an existing chart instead, check out [Embed a chart](./question.md).

## Embed the visual query builder

To let people build new questions with the visual query builder, use `new` as the question ID.

![Query builder](./images/embedded-query-builder.png)

As a web component:

```html
<metabase-question question-id="new"></metabase-question>
```

With the SDK:

```tsx
{% include_file "{{ dirname }}/sdk/snippets/questions/new-question.tsx" %}
```

> To narrow down what people can start from, list the entity types you want in the data picker with the `entity-types` attribute (web component) or the `entityTypes` prop (SDK). For example, `entity-types="['model']"` limits the picker to [models](../data-modeling/models.md).

Reviewer: Can we use a non-model example, seeing how we want to pretend that models don't exist :)

> Both query editors use the same `<metabase-question>` element as an embedded chart, so they take the same attributes and props. See the [Question reference](./question-reference.md).

Reviewer: what is "both" referring to here? we only talked about one query editor in this section. "Both visual query builder in SQL editor"?

## Embed the SQL editor

![Embedded SQL editor](./images/embedded-sql-editor.png)

To let people write native SQL, use `new-native` as the question ID.

As a web component:

```html
<metabase-question question-id="new-native"></metabase-question>
```

With the SDK:

```tsx
{% include_file "{{ dirname }}/sdk/snippets/questions/new-native-question.tsx" %}
```

> Everyone still queries through their own Metabase account, so people can only run SQL against databases their groups have permission to query. See [data permissions](../permissions/data.md).

Reviewer: This note needs to be in the previous section too

## Let people save questions

### Saving with web components

> With a web component, turn saving on with `is-save-enabled="true"`, and set the collection for new questions to land in with `target-collection`:

Reviewer: Target collection is optional, right? probably a good recommendation to always set it tho..
